tests/test_cli.py

```diff
@@ -36,8 +37,12 @@ def workdir(tmp_path):
 def execute(command, exit_code=0):
     runner = CliRunner()
     result = runner.invoke(cli, command)
-    assert result.exit_code == exit_code, result.output
-    return result.output
+    debugging = result.output
+    if exit_code == 0 and result.exit_code != 0 and result.exc_info:
+        _, _, tb = result.exc_info
+        debugging = traceback.print_tb(tb)
```

Comment:
This will return as follows: 
```
command = ['config', 'http://foo', '--profile', 'foo', '--config', '/private/var/folders/z9/tjd43wyd11j9h9tn51m84jjm0000gn/T/pytest-of-jeremy/pytest-14/test_command_config0/substra-cli/cli.cfg', ...], exit_code = 0

    def execute(command, exit_code=0):
        runner = CliRunner()
        result = runner.invoke(cli, command)
        debugging = result.output
        if exit_code == 0 and result.exit_code != 0 and result.exc_info:
            _, _, tb = result.exc_info
            debugging = traceback.print_tb(tb)
>       assert result.exit_code == exit_code, debugging
E       AssertionError: None
E       assert 1 == 0
E         -1
E         +0

tests/test_cli.py:44: AssertionError
----------------------------------------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------------------------------------
  File "/Users/jeremy/.virtualenvs/substra/lib/python3.7/site-packages/click/testing.py", line 326, in invoke
    cli.main(args=args or (), prog_name=prog_name, **extra)
  File "/Users/jeremy/.virtualenvs/substra/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/jeremy/.virtualenvs/substra/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/jeremy/.virtualenvs/substra/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/jeremy/.virtualenvs/substra/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/jeremy/substra/substra/substra/cli/interface.py", line 250, in add_profile_to_config
    raise Exception('nope')
```

You'll notice the `AssertionError: None`. This is because the print_tb doesn't return anything, it only prints the traceback. As a result, the traceback isn't next to the assertion error but in the `captured stderr call` section. If multiple test fail, all outputs will be grouped in this section, not really facilitating the debug (there is no way to know which test prompted which traceback).

Using format_tb doesn't solve the issue though, I guess in other to have a proper output you could have a look at pytest's "fail" function and pytest hooks (https://docs.pytest.org/en/latest/example/simple.html#post-process-test-reports-failures) I have no idea how to do it though, this is really about reading a lot of docs and trying out solutions.

Reply:
I can return `print_tb` in the `stdout` call section, by using this:
`debugging = traceback.print_tb(tb, file=sys.stdout)`.
But I still have the "AssertionError: None" indeed 🤔 

I'll try to use pytest.fail so 🙂